File: hackathon_CPP_SDK/src-gen/nevonex/gpsplugin/GPSPlugin.cpp
/*
 * nevonex/gpsplugin/GPSPlugin.cpp
 * Copyright (c) Robert Bosch GmbH. All rights reserved.
 */

#include "GPSPlugin.hpp"
#include <ecore/EObject.hpp>
#include <ecore/EClass.hpp>
#include "nevonex/gpsplugin/GpspluginPackage.hpp"
#include <ecorecpp/mapping.hpp>
#include <nevonex-fcal-platform/log/Logger.hpp>

/*PROTECTED REGION ID(GPSPlugin.cpp) START*/
// Please, enable the protected region if you add manually written code.
// To do this, add the keyword ENABLED before START.
/*PROTECTED REGION END*/

using namespace ::nevonex::gpsplugin;
using namespace ::nevonex::log;

// Default constructor
GPSPlugin::GPSPlugin() : m_gPSSensorPosition(0), m_internalGpsDetailedInfo(0)
{
    // Machine does NOT call _initialize() — Provider manages initialization
    m_interfaceDetailMap = {
    };

    /*PROTECTED REGION ID(GPSPlugin__GPSPlugin) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

GPSPlugin::~GPSPlugin()
{
    /*PROTECTED REGION ID(GPSPlugin__Destructor) START*/
    // Please, enable the protected region if you add manually written code.
    // To do this, add the keyword ENABLED before START.
    /*PROTECTED REGION END*/
}

// References
::ecore::EString GPSPlugin::getGPSSensorPosition() const
{
    return m_gPSSensorPosition;
}

void GPSPlugin::setGPSSensorPosition(::ecore::EString _gPSSensorPosition)
{
    ::ecore::EString _old_gPSSensorPosition = m_gPSSensorPosition;
    m_gPSSensorPosition = _gPSSensorPosition;
    SDK_FCAL_LOG(SeverityLevel::debug) << "GPSPlugin property \"gPSSensorPosition\" has changed.";
    notifyPropertyChange("gPSSensorPosition", _old_gPSSensorPosition, m_gPSSensorPosition);
}

::nevonex::fcal::INTERNAL_GPS_DETAILEDINFO_R_ptr GPSPlugin::getInternalGpsDetailedInfo() const
{
    return m_internalGpsDetailedInfo;
}

void GPSPlugin::setInternalGpsDetailedInfo(::nevonex::fcal::INTERNAL_GPS_DETAILEDINFO_R_ptr _internalGpsDetailedInfo)
{
    ::nevonex::fcal::INTERNAL_GPS_DETAILEDINFO_R_ptr _old_internalGpsDetailedInfo = m_internalGpsDetailedInfo;
    m_internalGpsDetailedInfo = _internalGpsDetailedInfo;
    SDK_FCAL_LOG(SeverityLevel::debug) << "GPSPlugin property \"internalGpsDetailedInfo\" has changed.";
    notifyPropertyChange("internalGpsDetailedInfo", _old_internalGpsDetailedInfo, m_internalGpsDetailedInfo);
}

// FCAL timestamp/valid
::ecore::ELong GPSPlugin::getGPSSensorPosition_Timestamp()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).getTimestamp();
    return 0;
}

::ecore::EBoolean GPSPlugin::isGPSSensorPosition_Valid()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).isValid();
    return false;
}

::ecore::ELong GPSPlugin::getInternalGpsDetailedInfo_Timestamp()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).getTimestamp();
    return 0;
}

::ecore::EBoolean GPSPlugin::isInternalGpsDetailedInfo_Valid()
{
    if (m_interfaceDetailMap.count(0))
        return m_interfaceDetailMap.at(0).isValid();
    return false;
}

// PropertyChange delegation (diamond inheritance requires explicit override)
void GPSPlugin::addPropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::addPropertyChangeListener(_listener);
}

void GPSPlugin::removePropertyChangeListener(
        ::nevonex::types::PropertyChangeListener _listener)
{
    using namespace ::nevonex::types;
    PropertyChange::removePropertyChangeListener(_listener);
}

void GPSPlugin::notifyPropertyChange(::ecore::EString const &_name,
        ::ecore::EJavaObject const &_oldValue,
        ::ecore::EJavaObject const &_newValue)
{
    using namespace ::nevonex::types;
    PropertyChange::notifyPropertyChange(_name, _oldValue, _newValue);
}

File: com.bosch.fsp.hackathon.gen/src/com/bosch/nevonex/gpsplugin/impl/GPSPlugin.java
/**
Copyright (c) Robert Bosch GmbH. All rights reserved.
*/
package com.bosch.nevonex.gpsplugin.impl;

import com.bosch.fsp.logger.FCALLogs;
import com.bosch.fsp.logger.LoggerConstants;

import com.bosch.fsp.platform.InterfaceDetails;

import com.bosch.fsp.runtime.feature.IMachine;

import com.bosch.nevonex.common.ITopicObject;
import com.bosch.nevonex.common.impl.TopicObject;

import com.bosch.nevonex.fcal.IINTERNAL_GPS_DETAILEDINFO_R;

import com.bosch.nevonex.gpsplugin.IGPSPlugin;

import com.bosch.nevonex.types.IPropertyChange;
import com.bosch.nevonex.types.impl.TypesPackage;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;

import java.lang.reflect.InvocationTargetException;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.eclipse.emf.common.util.BasicEList;
import org.eclipse.emf.common.util.EList;

import org.eclipse.emf.ecore.EClass;
import org.eclipse.emf.ecore.InternalEObject;

import org.eclipse.emf.ecore.util.EDataTypeUniqueEList;

/**
 * An implementation of the model object '<em><b>GPSPlugin</b></em>'.
 * @generated
 */
public class GPSPlugin extends TopicObject implements IGPSPlugin {
	protected EList<PropertyChangeListener> listeners;
	protected Map<String, InterfaceDetails> interfaceDetailMap = new HashMap<>();
	protected String gPSSensorPosition = null;
	protected IINTERNAL_GPS_DETAILEDINFO_R internalGpsDetailedInfo;

	/**
	 * @generated
	 */
	protected GPSPlugin() {
		super();
		initInterfaceDetails();
	}

	private void initInterfaceDetails() {
		// Interface details initialization
	}

	/**
	 * @generated
	 */
	@Override
	protected EClass eStaticClass() {
		return GpspluginPackage.Literals.GPS_PLUGIN;
	}

	public String getGPSSensorPosition() {
		return gPSSensorPosition;
	}

	public void setGPSSensorPosition(String newGPSSensorPosition) {
		gPSSensorPosition = newGPSSensorPosition;
	}

	/**
	 * @generated
	 */
	public IINTERNAL_GPS_DETAILEDINFO_R getInternalGpsDetailedInfo() {
		if (internalGpsDetailedInfo != null && internalGpsDetailedInfo.eIsProxy()) {
			InternalEObject oldInternalGpsDetailedInfo = (InternalEObject) internalGpsDetailedInfo;
			internalGpsDetailedInfo = (IINTERNAL_GPS_DETAILEDINFO_R) eResolveProxy(oldInternalGpsDetailedInfo);
			if (internalGpsDetailedInfo != oldInternalGpsDetailedInfo) {
			}
		}
		return internalGpsDetailedInfo;
	}

	/**
	 * @generated
	 */
	public IINTERNAL_GPS_DETAILEDINFO_R basicGetInternalGpsDetailedInfo() {
		return internalGpsDetailedInfo;
	}

	/**
	 * @generated
	 */
	public void setInternalGpsDetailedInfo(IINTERNAL_GPS_DETAILEDINFO_R newInternalGpsDetailedInfo) {
		IINTERNAL_GPS_DETAILEDINFO_R oldInternalGpsDetailedInfo = internalGpsDetailedInfo;
		internalGpsDetailedInfo = newInternalGpsDetailedInfo;
		FCALLogs.getInstance().log
				.debug(LoggerConstants.LOG_SDK_PREFIX + " GPSPlugin property \"internalGpsDetailedInfo\" changed from  "
						+ oldInternalGpsDetailedInfo + " to " + newInternalGpsDetailedInfo);
		notifyPropertyChange("internalGpsDetailedInfo", oldInternalGpsDetailedInfo, newInternalGpsDetailedInfo);
	}

	/**
	 * @generated
	 */
	public long getInternalGpsDetailedInfo_Timestamp() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("internalGpsDetailedInfo");
		if (interfaceDetails != null) {
			return interfaceDetails.getTimestmap();
		}
		return -1;
	}

	/**
	 * @generated
	 */
	public boolean isInternalGpsDetailedInfo_Valid() {
		InterfaceDetails interfaceDetails = interfaceDetailMap.get("internalGpsDetailedInfo");
		if (interfaceDetails != null) {
			return interfaceDetails.isValid();
		}
		return false;
	}

	/**
	 * @generated
	 */
	public void addPropertyChangeListener(PropertyChangeListener listener) {
		if (listeners == null) {
			listeners = new BasicEList<>();
		}
		listeners.add(listener);
	}

	/**
	 * @generated
	 */
	public void removePropertyChangeListener(PropertyChangeListener listener) {
		if (listeners != null) {
			listeners.remove(listener);
		}
	}

	/**
	 * @generated
	 */
	public void notifyPropertyChange(String name, Object oldValue, Object newValue) {
		if (listeners != null) {
			for (PropertyChangeListener listener : this.listeners) {
				listener.propertyChange(new PropertyChangeEvent(this, name, oldValue, newValue));
			}
		}
	}

	/**
	 * @generated
	 */
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
		case GpspluginPackage.IGPS_PLUGIN__GPS_SENSOR_POSITION:
			return getGPSSensorPosition();
		case GpspluginPackage.IGPS_PLUGIN__INTERNAL_GPS_DETAILED_INFO:
			if (resolve)
				return getInternalGpsDetailedInfo();
			return basicGetInternalGpsDetailedInfo();
		}
		return super.eGet(featureID, resolve, coreType);
	}

	/**
	 * @generated
	 */
	@SuppressWarnings("unchecked")
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
		case GpspluginPackage.IGPS_PLUGIN__GPS_SENSOR_POSITION:
			setGPSSensorPosition((String) newValue);
			return;
		case GpspluginPackage.IGPS_PLUGIN__INTERNAL_GPS_DETAILED_INFO:
			setInternalGpsDetailedInfo((IINTERNAL_GPS_DETAILEDINFO_R) newValue);
			return;
		}
		super.eSet(featureID, newValue);
	}

	/**
	 * @generated
	 */
	@Override
	public void eUnset(int featureID) {
		switch (featureID) {
		case GpspluginPackage.IGPS_PLUGIN__GPS_SENSOR_POSITION:
			setGPSSensorPosition(null);
			return;
		case GpspluginPackage.IGPS_PLUGIN__INTERNAL_GPS_DETAILED_INFO:
			setInternalGpsDetailedInfo((IINTERNAL_GPS_DETAILEDINFO_R) null);
			return;
		}
		super.eUnset(featureID);
	}

	/**
	 * @generated
	 */
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
		case GpspluginPackage.IGPS_PLUGIN__GPS_SENSOR_POSITION:
			return gPSSensorPosition != null;
		case GpspluginPackage.IGPS_PLUGIN__INTERNAL_GPS_DETAILED_INFO:
			return internalGpsDetailedInfo != null;
		}
		return super.eIsSet(featureID);
	}
} //GPSPlugin
